staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go

```diff
@@ -0,0 +1,345 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validators
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/gengo/v2/types"
+)
+
+const (
+	listTypeTag   = "k8s:listType2"
+	listMapKeyTag = "k8s:listMapKey2"
+	eachValTag    = "k8s:eachVal2"
+)
+
+func init() {
+	// Lists with list-map semantics are comprised of multiple tags, which need
+	// to share information between them.
+	shared := map[string]*listMap{} // keyed by the fieldpath
+	RegisterTagValidator(listTypeTagValidator{shared})
+	RegisterTagValidator(listMapKeyTagValidator{shared})
+	RegisterTagValidator(&eachValTagValidator{shared, nil})
+}
+
+// listMap collects information about a single list with map semantics.
+type listMap struct {
+	declaredAsMap bool
+	keyFields     []string
+}
+
+type listTypeTagValidator struct {
+	byFieldPath map[string]*listMap
+}
+
+func (listTypeTagValidator) Init(Config) {}
+
+func (listTypeTagValidator) TagName() string {
+	return listTypeTag
+}
+
+func (listTypeTagValidator) ValidScopes() sets.Set[Scope] {
+	return listTagsValidScopes
+}
+
+func (lttv listTypeTagValidator) GetValidations(context Context, _ []string, payload string) (Validations, error) {
+	t := context.Type
+	if t.Kind == types.Alias {
+		t = t.Underlying
+	}
+	if t.Kind != types.Slice && t.Kind != types.Array {
+		return Validations{}, fmt.Errorf("can only be used on list types")
+	}
+
+	if payload == "map" {
+		if realType(t.Elem).Kind != types.Struct {
+			return Validations{}, fmt.Errorf("only lists of structs can be list-maps")
+		}
+
+		// Save the fact that this list is a map.
+		if lttv.byFieldPath[context.Path.String()] == nil {
+			lttv.byFieldPath[context.Path.String()] = &listMap{}
+		}
+		lm := lttv.byFieldPath[context.Path.String()]
+		lm.declaredAsMap = true
+	} // FIXME:else?
```

Comment:
Supported values: atomic/set/map

For "atomic" and "set" not additional validation or tracking of information is required.

If the payload is a unrecognized string, we should report and error I think.

Reply:
So say we all.  In next push.